dotnet/samples/Concepts/Optimization/PluginSelection.cs

```diff
@@ -0,0 +1,271 @@
+        kernel.ImportPluginFromType<WeatherPlugin>();
+        kernel.ImportPluginFromType<EmailPlugin>();
+        kernel.ImportPluginFromType<NewsPlugin>();
+        kernel.ImportPluginFromType<CalendarPlugin>();
+
+        // Enable automatic function calling.
+        var executionSettings = new OpenAIPromptExecutionSettings { ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions };
+
+        // Define kernel arguments with specific request.
+        var kernelArguments = new KernelArguments(executionSettings) { ["Request"] = "Provide latest headlines" };
+
+        // Invoke the request without plugin selection filter first for comparison purposes.
+        Console.WriteLine("Run without filter:");
+        var result = await kernel.InvokePromptAsync("{{$Request}}", kernelArguments);
+
+        Console.WriteLine(result);
+        Console.WriteLine(result.Metadata?["Usage"]?.AsJson()); // All functions were shared with AI. Total tokens: ~250
+
+        // Define plugin selection filter.
+        var filter = new PluginSelectionFilter(
+            kernel.GetRequiredService<IMemoryStore>(),
+            kernel.GetRequiredService<ITextEmbeddingGenerationService>(),
+            kernel.GetRequiredService<ILogger>(),
+            numberOfBestFunctions: 1);
+
+        // Add filter to kernel.
+        kernel.FunctionInvocationFilters.Add(filter);
+
+        // Invoke the request with plugin selection filter.
+        Console.WriteLine("\nRun with filter:");
+        result = await kernel.InvokePromptAsync("{{$Request}}", kernelArguments);
+
+        Console.WriteLine(result);
+        Console.WriteLine(result.Metadata?["Usage"]?.AsJson()); // Just one function was shared with AI. Total tokens: ~150
+    }
+
+    /// <summary>
+    /// Filter which performs vector similarity search on imported functions in <see cref="Kernel"/>
+    /// to select the best ones to share with AI.
+    /// </summary>
+    private sealed class PluginSelectionFilter(
+        IMemoryStore memoryStore,
+        ITextEmbeddingGenerationService textEmbeddingGenerationService,
+        ILogger logger,
+        int numberOfBestFunctions) : IFunctionInvocationFilter
+    {
+        /// <summary>Collection name to use in memory store.</summary>
+        private const string CollectionName = "functions";
+
+        public async Task OnFunctionInvocationAsync(FunctionInvocationContext context, Func<FunctionInvocationContext, Task> next)
+        {
+            var request = GetRequestArgument(context.Arguments);
+
+            // Execute plugin selection logic for "InvokePrompt" function only, as main entry point.
+            if (context.Function.Name.Contains(nameof(KernelExtensions.InvokePromptAsync)) && !string.IsNullOrWhiteSpace(request))
+            {
+                // Get imported plugins in kernel.
+                var plugins = context.Kernel.Plugins;
+
+                // Store plugins in memory store.
+                await this.StorePluginsAsync(plugins);
```

Comment:
We probably want to do this outside of the filter (maybe even outside of the service and we can maybe add a comment suggesting that), since generating the embeddings again for each request will probably negate the cost reduction we get from not passing all functions to the LLM. It will also add a lot of latency.

Reply:
That's a good point. Initially I thought that maybe for this example it would be enough to place that logic inside filter, but I think it's better to show the right patterns from scratch. I moved plugins store operation outside the filter to perform it only once.